apps/web/utils/discord-markdown.ts

```diff
@@ -3,77 +3,158 @@ import { load } from 'cheerio'
-  const postIds = new Set<string>(Array.from(content.matchAll(regex), (m) => m.groups?.channel ?? ''))
-
-  const posts =
-    postIds.size > 0
-      ? await db
-        .selectFrom('posts')
-        .select(['snowflakeId', 'title'])
-        .where('snowflakeId', 'in', Array.from(postIds))
-        .execute()
-      : []
-
-  // Parse the content again, this time replacing the link
-  content = content.replace(regex, (match, guild, channel, _, message) => {
-    const post = posts.find((p) => p.snowflakeId === channel)
+
+interface UserCache {
+  snowflakeId: string
+  username: string
+}
+
+interface ChannelCache {
+  snowflakeId: string
+  name: string
+}
+
+interface PostCache {
+  snowflakeId: string
+  title: string
+}
+
+const QUERY_REVALIDATE_TIME = 60 // 1 minute
+
+const fetchUser = (userId: string) => unstable_cache(async () => {
+  return (
+    await db
+      .selectFrom('users')
+      .select(['snowflakeId', 'username'])
+      .where('snowflakeId', '=', userId)
+      .executeTakeFirst()
+  ) || null
+
+}, [userId], { revalidate: QUERY_REVALIDATE_TIME })()
+
+const fetchChannel = (channelId: string) => unstable_cache(async () => {
+  return (
+    await db
+      .selectFrom('channels')
+      .select(['snowflakeId', 'name'])
+      .where('snowflakeId', '=', channelId)
+      .executeTakeFirst()
+  ) || null
+
+}, [channelId], { revalidate: QUERY_REVALIDATE_TIME })()
+
+const fetchPost = (postId: string) => unstable_cache(async () => {
+  return (
+    await db
+      .selectFrom('posts')
+      .select(['snowflakeId', 'title'])
+      .where('snowflakeId', '=', postId)
+      .executeTakeFirst()
+  ) || null
+
+}, [postId], { revalidate: QUERY_REVALIDATE_TIME })()
+
+const channelLinkRegex = /https:\/\/discord\.com\/channels\/(?<guild>\d+)\/(?<channel>\d+)(\/(?<message>\d+))?/g;
+const userMention = /<@!?(?<user>\d+)>/g;
+const channelMention = /<#(?<channel>\d+)>/g;
+
+export const extractMentions = (content: string) => {
+  const postIds = new Set<string>(Array.from(content.matchAll(channelLinkRegex), (m) => m.groups?.channel ?? ''))
+  const memberIds = new Set<string>(Array.from(content.matchAll(userMention), (m) => m.groups?.user ?? ''))
+  const channelIds = new Set<string>(Array.from(content.matchAll(channelMention), (m) => m.groups?.channel ?? ''))
+  return { postIds, memberIds, channelIds }
+}
+
+export const fetchMentions = async (content: string) => {
+  const { postIds, memberIds, channelIds } = extractMentions(content)
+
+  // Fetch from db/cache
+  const [posts, users, channels] = (await Promise.all([
+    Promise.all(Array.from(postIds).map(fetchPost)),
+    Promise.all(Array.from(memberIds).map(fetchUser)),
+    Promise.all(Array.from(channelIds).map(fetchChannel)),
+  ]))
+
+  // Filter out null values
+  const postsFiltered = posts.filter((p) => p) as PostCache[]
+  const usersFiltered = users.filter((u) => u) as UserCache[]
+  const channelsFiltered = channels.filter((c) => c) as ChannelCache[]
```

Comment:
We probably don't need this, it is fine to have the arrays with possible `undefined` values because we check for them anyway later on

Reply:
Same as above, but the reason it needed `null` was because the filter didn't like filtering undefined variables.